packages/core/src/oidc/grants/token-exchange/account.ts

```diff
@@ -1,35 +1,144 @@
+    ) => Promise<KeyLike | Uint8Array>;
+    issuer: string;
+  };
+};
+
+/**
+ * Validates an impersonation token (subject token created via impersonation API).
+ * These tokens start with "sub_" prefix and are stored in the database.
+ */
+const validateImpersonationToken = async (
+  findSubjectToken: Queries['subjectTokens']['findSubjectToken'],
+  subjectToken: string
+): Promise<{ userId: string; subjectTokenId: string }> => {
+  const token = await trySafe(async () => findSubjectToken(subjectToken));
+  assertThat(token, new InvalidGrant('subject token not found'));
+  assertThat(token.expiresAt > Date.now(), new InvalidGrant('subject token is expired'));
+  assertThat(!token.consumedAt, new InvalidGrant('subject token is already consumed'));
+
+  return {
+    userId: token.userId,
+    subjectTokenId: token.id,
+  };
+};
+
+/**
+ * Validates a JWT access token using the issuer's JWK set.
+ * JWT access tokens can be exchanged multiple times (not consumption-tracked).
+ */
+const validateJwtAccessToken = async (
   subjectToken: string,
-  type: string
-): Promise<{ userId: string; subjectTokenId?: string }> => {
+  jwtVerificationOptions: NonNullable<ValidateSubjectTokenParams['jwtVerificationOptions']>
+): Promise<{ userId: string }> => {
+  const { localJWKSet, issuer } = jwtVerificationOptions;
+
+  try {
+    const { payload } = await jwtVerify(subjectToken, localJWKSet, { issuer });
+    assertThat(payload.sub, new InvalidGrant('subject token does not contain a valid `sub` claim'));
+
+    // JWT access tokens are not consumption-tracked, so no subjectTokenId is returned.
+    // This allows the same token to be exchanged multiple times (e.g., by different services).
+    return { userId: payload.sub };
+  } catch (error) {
+    if (error instanceof errors.OIDCProviderError) {
+      throw error;
+    }
+    throw new InvalidGrant('invalid subject token');
+  }
+};
+
+/**
+ * Validates an opaque access token by looking it up via oidc-provider's AccessToken.find.
+ * Opaque access tokens can be exchanged multiple times (not consumption-tracked).
+ */
+const validateOpaqueAccessToken = async (
+  subjectToken: string,
+  AccessToken: Provider['AccessToken']
+): Promise<{ userId: string } | undefined> => {
+  // eslint-disable-next-line unicorn/no-array-callback-reference -- AccessToken.find is not an array method
+  const token = await trySafe(async () => AccessToken.find(subjectToken));
+  if (!token?.accountId) {
+    return;
+  }
+
+  // Check if the token is expired
+  if (token.isExpired) {
+    throw new InvalidGrant('subject token is expired');
+  }
+
+  // Opaque access tokens are not consumption-tracked, so no subjectTokenId is returned.
+  // This allows the same token to be exchanged multiple times (e.g., by different services).
+  return { userId: token.accountId };
+};
+
+/** Prefix used by impersonation tokens for backward compatibility detection. */
+const legacyImpersonationTokenPrefix = 'sub_';
```

Comment:
nit: 
```suggestion
const`IMPERSONATION_TOKEN_PREFIX` = 'sub_';
```
1. Remove the `legacy`. We always use `sub_` as the impersonation token's prefix. 
2. Use upper case for static constant.